sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,1632 @@
+            else:
+                bounds.append(getattr(self, var_name + "_bounds"))
+        return np.log(np.vstack(bounds))
+
+    @bounds.setter
+    def bounds(self, bounds):
+        """Sets the bounds on the kernel's hyperparameters theta.
+
+        Parameters
+        ----------
+        bounds : array, shape (n_dims, 2)
+            The bounds on the kernel's hyperparameters theta
+        """
+        i = 0
+        for var_name in self.theta_vars:
+            if not isinstance(var_name, six.string_types):
+                # vector-valued parameter
+                var_name, var_length = var_name
+                setattr(self, var_name + "_bounds",
+                        np.exp(bounds[i:i + var_length]))
+                i += var_length
+            else:
+                setattr(self, var_name + "_bounds", np.exp(bounds[i]))
+                i += 1
+
+        if i != len(bounds):
+            raise ValueError("bounds has not the correct number of entries."
+                             " Should be %d; given are %d"
+                             % (i, len(bounds)))
+
+    def __add__(self, b):
+        if not isinstance(b, Kernel):
+            return Sum(self, ConstantKernel(b))
+        return Sum(self, b)
+
+    def __radd__(self, b):
+        if not isinstance(b, Kernel):
+            return Sum(ConstantKernel(b), self)
+        return Sum(b, self)
+
+    def __mul__(self, b):
+        if not isinstance(b, Kernel):
+            return Product(self, ConstantKernel(b))
+        return Product(self, b)
+
+    def __rmul__(self, b):
+        if not isinstance(b, Kernel):
+            return Product(ConstantKernel(b), self)
+        return Product(b, self)
+
+    def __pow__(self, b):
+        return Exponentiation(self, b)
+
+    def __eq__(self, b):
+        if type(self) != type(b):
+            return False
+        params_a = self.get_params()
+        params_b = b.get_params()
+        for key in set(list(params_a.keys()) + list(params_b.keys())):
+            if np.any(params_a.get(key, None) != params_b.get(key, None)):
+                return False
+        return True
+
+    def __repr__(self):
+        return "{0}({1})".format(self.__class__.__name__,
+                                 ", ".join(map("{0:.3g}".format, self.theta)))
+
+    @abstractmethod
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """Evaluate the kernel."""
+
+    def diag(self, X):
+        """Returns the diagonal of the kernel k(X, X).
+
+        The result of this method is identical to np.diag(self(X)); however,
+        it can be evaluated more efficiently since only the diagonal is
+        evaluated.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples_X, n_features)
+            Left argument of the returned kernel k(X, Y)
+
+        Returns
+        -------
+        K_diag : array, shape (n_samples_X,)
+            Diagonal of kernel k(X, X)
+        """
+        return np.ones(X.shape[0])
```

Comment:
This doesn't seem to generally hold. So having it in the top class is dangerous, because it needs to be overwritten in kernels where this is not the case.